File: api/web/mappers/message_mapper.py
# ABOUTME: Message mapper converting between ClientMessage DTOs and domain Message entities
# ABOUTME: Handles tool invocations and attachments, expanding single ClientMessage to multiple Messages
import json
from typing import List
from ...domain.entities.message import Message, MessageRole, Attachment
from ...domain.entities.tool import ToolCall
from ..dtos.client_messages import ClientMessage, ToolInvocation, ToolInvocationState


class MessageMapper:
    """Maps between web DTOs and domain entities"""

    @staticmethod
    def to_domain(client_messages: List[ClientMessage]) -> List[Message]:
        """
        Convert ClientMessage list to domain Message list.

        May return more messages than input due to tool invocations being expanded.
        """
        messages = []

        for client_message in client_messages:
            # Main message with attachments
            attachments = []
            if client_message.experimental_attachments:
                for att in client_message.experimental_attachments:
                    attachments.append(Attachment(
                        content_type=att.contentType,
                        url=att.url
                    ))

            # Extract tool calls from CALL state invocations (for assistant messages)
            tool_calls = None
            if client_message.toolInvocations and client_message.role == "assistant":
                tool_calls = []
                for invocation in client_message.toolInvocations:
                    # Create tool call from either CALL or RESULT state
                    if invocation.state in [ToolInvocationState.CALL, ToolInvocationState.RESULT]:
                        tool_calls.append(ToolCall(
                            id=invocation.toolCallId,
                            name=invocation.toolName,
                            arguments=invocation.args
                        ))

            # Add main message
            messages.append(Message(
                role=MessageRole(client_message.role),
                content=client_message.content,
                attachments=attachments,
                tool_calls=tool_calls if tool_calls else None
            ))

            # Add tool result messages if present
            if client_message.toolInvocations:
                for invocation in client_message.toolInvocations:
                    if invocation.state == ToolInvocationState.RESULT:
                        messages.append(Message(
                            role=MessageRole.TOOL,
                            content=json.dumps(invocation.result),
                            tool_call_id=invocation.toolCallId
                        ))

        return messages

File: src/application/mappers/MessageMapper.ts
// ABOUTME: Maps between domain Message entities and DTOs
// ABOUTME: Handles conversion logic and data transformation

import { Message } from '../../domain/entities/Message';
import { ToolInvocation, ToolInvocationState } from '../../domain/entities/ToolInvocation';
import { MessageRole } from '../../domain/value-objects/MessageRole';
import { MessageContent } from '../../domain/value-objects/MessageContent';
import { Attachment } from '../../domain/value-objects/Attachment';
import { ToolName } from '../../domain/value-objects/ToolName';
import { MessageDto, ClientMessageDto, ToolInvocationDto, AttachmentDto } from '../dto/MessageDto';

export class MessageMapper {
  /**
   * Converts domain Message to DTO
   */
  static toDto(message: Message): MessageDto {
    return {
      id: message.getId(),
      role: message.getRole().getValue(),
      content: message.getContent().getValue(),
      attachments: message.getAttachments().map(a => this.attachmentToDto(a)),
      toolInvocations: message.getToolInvocations().map(t => this.toolInvocationToDto(t)),
      timestamp: message.getTimestamp().toISOString(),
    };
  }

  /**
   * Converts DTO to domain Message
   */
  static fromDto(dto: MessageDto): Message {
    const role = MessageRole.from(dto.role);
    const content = MessageContent.from(dto.content);

    const attachments = (dto.attachments || []).map(a =>
      Attachment.create(a.name, a.contentType, a.url)
    );

    const toolInvocations = (dto.toolInvocations || []).map(t =>
      ToolInvocation.create(
        t.toolCallId,
        ToolName.from(t.toolName),
        t.args
      )
    );

    if (dto.id) {
      return Message.createWithId(dto.id, role, content, attachments, toolInvocations);
    }
    return Message.create(role, content, attachments, toolInvocations);
  }

  /**
   * Converts ClientMessageDto to domain Message
   */
  static fromClientDto(dto: ClientMessageDto): Message {
    const role = MessageRole.from(dto.role);
    const content = MessageContent.from(dto.content);

    const attachments = (dto.experimental_attachments || []).map(a =>
      Attachment.create(a.name, a.contentType, a.url)
    );

    const toolInvocations = (dto.toolInvocations || []).map(t =>
      ToolInvocation.create(
        t.toolCallId,
        ToolName.from(t.toolName),
        t.args
      )
    );

    return Message.create(role, content, attachments, toolInvocations);
  }

  /**
   * Converts array of domain Messages to DTOs
   */
  static toDtoArray(messages: readonly Message[]): MessageDto[] {
    return messages.map(m => this.toDto(m));
  }

  /**
   * Converts array of DTOs to domain Messages
   */
  static fromDtoArray(dtos: MessageDto[]): Message[] {
    return dtos.map(dto => this.fromDto(dto));
  }

  /**
   * Converts array of ClientMessageDtos to domain Messages
   */
  static fromClientDtoArray(dtos: ClientMessageDto[]): Message[] {
    return dtos.map(dto => this.fromClientDto(dto));
  }

  private static attachmentToDto(attachment: Attachment): AttachmentDto {
    return attachment.toObject();
  }

  private static toolInvocationToDto(toolInvocation: ToolInvocation): ToolInvocationDto {
    const state = toolInvocation.getState();
    let dtoState: 'call' | 'partial-call' | 'result';

    if (state === ToolInvocationState.COMPLETED) {
      dtoState = 'result';
    } else if (state === ToolInvocationState.EXECUTING) {
      dtoState = 'partial-call';
    } else {
      dtoState = 'call';
    }

    return {
      state: dtoState,
      toolCallId: toolInvocation.getCallId(),
      toolName: toolInvocation.getToolName().getValue(),
      args: toolInvocation.getArgs(),
      result: toolInvocation.getResult(),
    };
  }
}